OrbitCore/LinuxPerfEventProcessor.h

```diff
@@ -46,23 +50,40 @@ class LinuxPerfEventProcessor {
       std::unique_ptr<LinuxPerfEventVisitor> visitor)
       : visitor_(std::move(visitor)) {}
 
-  void Push(std::unique_ptr<LinuxPerfEvent> event);
+  void PushEvent(int origin_fd, std::unique_ptr<LinuxPerfEvent> event);
 
   void ProcessAllEvents();
 
   void ProcessOldEvents();
 
  private:
-  // TODO: Inserting into this buffer is logarithmic in its size. However, all
-  //  events within one ring buffer are already sorted. We could instead
-  //  implement a priority queue of ring buffers and be logarithmic in the
-  //  number of buffers. This would require moving a ring buffer down the heap
-  //  once an event has been processed, by removing-and-readding or with a
-  //  custom priority queue.
-  std::priority_queue<std::unique_ptr<LinuxPerfEvent>,
-                      std::vector<std::unique_ptr<LinuxPerfEvent>>,
-                      TimestampReverseCompare>
-      event_queue_;
+  bool HasEvent();
+  LinuxPerfEvent* TopEvent();
+  std::unique_ptr<LinuxPerfEvent> PopEvent();
+
+  // Instead of keeping a single priority queue with all the events to process,
+  // on which push/pop operations would be logarithmic in the number of events,
+  // we leverage the fact that events coming from the same perf_event_open ring
+  // buffer are already sorted. We then keep a priority queue of queues, where
+  // the events in each queue come from the same file descriptor. Whenever an
+  // event is removed from a queue, we need to move such queue down the priority
+  // queue. As std::priority_queue does not support decreasing the priority of
+  // an element, we achieve this by removing and re-inserting.
+  // In order to be able to add an event to a queue, we also need to maintain
+  // the association between a queue and its file descriptor. This is what the
+  // map is for.
```

Comment:
sounds like this is something that could be extracted to a separate class? It would also allow to add a test for the reinsertion part.. so that it could be easily tested after addressing todo below..

Reply:
Good point.